File: iraf/sys/osb/bswap2.f
c BSWAP2 - Move bytes from array "a" to array "b", swapping successive
c pairs of bytes.

	subroutine bswap2 (a, aoff, b, boff, nbytes)

	character*1 a(*), b(*), temp
	integer	aoff, boff, nbytes, i
	integer aoff1, boff1

		aoff1 = aoff + 1
		boff1 = boff + 1

 		do 10 i = 0, nbytes-1, 2
		    temp = a(aoff1+i)
		    if (i .ne. nbytes) then
			b(boff1+i) = a(aoff+i)
		    endif
		    b(boff+i) = temp
 10		continue
	end

File: iraf/sys/osb/bswap2.c
/* Copyright(c) 1986 Association of Universities for Research in Astronomy Inc.
 */

#define import_spp
#define import_knames
#include <iraf.h>

/* BSWAP2 - Move bytes from array "a" to array "b", swapping successive
 * pairs of bytes.  The two arrays may be the same but may not be offset
 * and overlapping.
 */
/* a      : input array			*/
/* aoff   : first byte in input array	*/
/* b      : output array		*/
/* boff   : first byte in output array	*/
/* nbytes : number of bytes to swap	*/
int BSWAP2 ( void *a, XINT *aoff, void *b, XINT *boff, XINT *nbytes )
{
	char *ip, *op, *otop;
	unsigned int temp;

	ip = (char *)a + *aoff - 1;
	op = (char *)b + *boff - 1;
	otop = op + (*nbytes & ~1);

	/* Swap successive pairs of bytes.
	 */
	while (op < otop) {
	    temp  = *ip++;
	    *op++ = *ip++;
	    *op++ = temp;
	}

	/* If there is an odd byte left, move it to the output array.
	 */
	if (*nbytes & 1)
	    *op = *ip;

	return 0;
}
